sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java

```diff
@@ -353,4 +354,28 @@ public Long create(PipelineOptions options) {
       return NEXT_ID.getAndIncrement();
     }
   }
+
+  /**
+   * A user agent string describing the pipeline to external services.
+   *
+   * <p>The string defaults to "[name]/[version]" based on the properties of the Beam release.
+   */
+  @Description("A user agent string describing the pipeline to external services. "
+      + "The default name is {@code [name]/[version]}"
+      + " where name and version are properties of the Apache Beam release.")
+  @Default.InstanceFactory(UserAgentFactory.class)
+  String getUserAgent();
+  void setUserAgent(String userAgent);
+
+  /**
+   * Returns a user agent string constructed from {@link ReleaseInfo#getName()} and
+   * {@link ReleaseInfo#getVersion()}, in the format "[name]/[version]".
+   */
+  class UserAgentFactory implements DefaultValueFactory<String> {
+    @Override
+    public String create(PipelineOptions options) {
+      ReleaseInfo info = ReleaseInfo.getReleaseInfo();
+      return String.format("%s/%s", info.getName(), info.getVersion()).replace(" ", "_");
```

Comment:
Is there a reason why we replace `" "` with `"_"`?

Reply:
That's how BigtableIO originally had it when reading from ReleaseInfo. Why it did that replacement I'm not sure.

Reply:
It turns out they are following https://www.ietf.org/rfc/rfc2616.txt where:
```
user agent         = 1*(product | comment)
product            = token ["/" product-version]
product-version    = token
```

Add to the user agent comment that this field should represent a user agent as per RFC2616 and give the BNF form that I describe above. A token is limited to any characters not including a separator.